### PR TITLE
feat: implement EventLoopGroup on DefaultEventLoopGroup

### LINKED ISSUE #33: Implement DefaultEventLoopGroup
This is about implementing `DefaultEventLoopGroup` which implements `EventLoopGroup` interface on core package

### PR DESCRIPTION
**Pull request by raspLarry**
- resolve #33

### CODE REVIEW

File: transport/src/main/java/io/el/channel/DefaultChannelEventLoop.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+/**
+ * TODO: implement this
+ */
+public class DefaultChannelEventLoop implements ChannelEventLoop {
```
Comment (zeroFruit) [inline transport/src/main/java/io/el/channel/DefaultChannelEventLoop.java:25]: Implementation of this class is not included in issue #37

File: transport/src/main/java/io/el/channel/DefaultChannelEventLoopGroup.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+public class DefaultChannelEventLoopGroup implements ChannelEventLoopGroup {
```
Comment (zeroFruit) [inline transport/src/main/java/io/el/channel/DefaultChannelEventLoopGroup.java:18]: Implementation of this class is not included in issue https://github.com/zeroFruit/el/pull/37

File: transport/src/main/java/io/el/channel/DefaultChannelPromise.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import io.el.concurrent.Promise;
+import io.el.concurrent.PromiseListener;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * TODO: implement this
+ */
+public class DefaultChannelPromise implements ChannelPromise {
```
Comment (zeroFruit) [inline transport/src/main/java/io/el/channel/DefaultChannelPromise.java:12]: Implementation of this class is not included in issue https://github.com/zeroFruit/el/pull/37

File: core/src/main/java/io/el/concurrent/DefaultEventLoopGroup.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+  /***
+   * Create {@link EventLoop} children with size of {@param nThreads},
+   * then add chooser with {@param chooserFactory}
+   */
+  protected DefaultEventLoopGroup(
+      int nThreads,
+      Executor executor,
+      EventLoopChooserFactory chooserFactory
+  ) {
+    checkPositive(nThreads, "nThreads");
```
Comment (majecty) [inline core/src/main/java/io/el/concurrent/DefaultEventLoopGroup.java:34]: :+1:

File: core/src/main/java/io/el/concurrent/DefaultEventLoopGroup.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    for (int i = 0; i < nThreads; i++) {
+      try {
+        this.children.add(this.newChild(executor));
+      } catch (Exception e) {
+        for (int j = 0; j < i; j++) {
+          if (this.children.get(j).shutdownGracefully(DEFAULT_SHUTDOWN_TIMEOUT, TimeUnit.SECONDS)) {
+            EventLoop el = this.children.get(j);
+            try {
+              while (!el.isTerminated()) {
+                el.awaitTermination(Integer.MAX_VALUE, TimeUnit.SECONDS);
```
Comment (majecty) [inline core/src/main/java/io/el/concurrent/DefaultEventLoopGroup.java:50]: The default termination time seems too big. How about waiting few seconds?

File: core/src/main/java/io/el/concurrent/DefaultEventLoopGroup.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    return this.children.iterator();
+  }
+
+  @Override
+  public boolean shutdownGracefully(long timeout, TimeUnit unit) {
+    return this.children.stream()
+        .map(c -> c.shutdownGracefully(timeout, unit))
+        .filter(s -> s.equals(false))
+        .findAny()
+        .orElse(true);
```
Comment (majecty) [inline core/src/main/java/io/el/concurrent/DefaultEventLoopGroup.java:85]: It looks difficult to understand for me. You don't need to fix it.

File: core/src/main/java/io/el/concurrent/DefaultEventLoopGroup.java [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+          if (this.children.get(j).shutdownGracefully(DEFAULT_SHUTDOWN_TIMEOUT, TimeUnit.SECONDS)) {
+            EventLoop el = this.children.get(j);
+            try {
+              while (!el.isTerminated()) {
+                el.awaitTermination(Integer.MAX_VALUE, TimeUnit.SECONDS);
+              }
+            } catch (InterruptedException interrupted) {
+              Thread.currentThread().interrupt();
+              break;
+            }
```
Comment (zeroFruit) [inline core/src/main/java/io/el/concurrent/DefaultEventLoopGroup.java:55]: I think it's better to call `shutdownGracefully()` for all child first then `awaitTermination()` all. Because `shutdownGracefully()` is done in asynchronous way. So call all child shutdown then await all.

File: core/src/main/java/io/el/concurrent/DefaultEventLoopGroup.java [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+
+  @Override
+  public void shutdown() {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public List<Runnable> shutdownNow() {
+    throw new UnsupportedOperationException();
+  }
```
Comment (zeroFruit) [inline core/src/main/java/io/el/concurrent/DefaultEventLoopGroup.java:96]: These not supported methods should be located in AbstractEventLoop

File: core/src/main/java/io/el/concurrent/DefaultEventLoopGroup.java [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    return this.next().submit(task, result);
+  }
+
+  @Override
+  public Promise<?> submit(Runnable task) {
+    return this.next().submit(task);
+  }
+
+  @Override
+  public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks)
```
Comment (zeroFruit) [inline core/src/main/java/io/el/concurrent/DefaultEventLoopGroup.java:159]: All invoke series methods are from `ExecutorService` these will not be used I think. So let's move these methods to AbstractEventLoop and throw UnsupportedException

File: core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+  };
+  private final EventLoopChooserFactory mockEventLoopChooserFactory = mock(
+      EventLoopChooserFactory.class);
+
+  @BeforeEach
+  public void setup() throws Exception {
+    when(mockEventLoopChooserFactory.newChooser(any())).thenReturn(chooser);
+  }
+
+  @Test
```
Comment (zeroFruit) [inline core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java:53]: Please change the test format following the convention

File: core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+  void testAwaitTermination() throws InterruptedException {
+    boolean awaitTermination = false;
+    try {
+      defaultEventLoopGroup.execute(mockTask);
+      awaitTermination = defaultEventLoopGroup.awaitTermination(1, TimeUnit.SECONDS);
+    } finally {
+      assertTrue(awaitTermination);
+      assertTrue(defaultEventLoopGroup.next().isTerminated());
+    }
+  }
```
Comment (zeroFruit) [inline core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java:86]: I think these simple test case is not need. I think it's better to test constructor

File: core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+      return null;
+    }
+  }
+
+  @Nested
+  @DisplayName("On DefaultEventLoopGroup constructor")
+  class DefaultEventLoopGroupConstructor {
+
+    @Test
+    @DisplayName("When construct with proper newChild() method, then shutdownGracefully and shutdown")
```
Comment (zeroFruit) [inline core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java:207]: plz update `@DisplayName` contents

File: core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+      try {
+        new MockFailedDefaultEventLoopGroup(
+            3,
+            mockExecutor,
+            eventLoopChooserFactory
+        );
+      } catch (IllegalStateException e) {
+        assertFalse(defaultEventLoopGroup.isTerminated());
+      }
+    }
```
Comment (zeroFruit) [inline core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java:218]: Is this a success case?

File: core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+
+    @Override
+    protected EventLoop newChild(Executor executor) {
+      try {
+        throw new Exception();
+      } catch (Exception e) {
+
+      }
+      return null;
```
Comment (zeroFruit) [inline core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java:198]: why don't just
```java
throw new Exception();
```
instead of
```java
try {
  throw new Exception();
} catch (Exception e) {

}
```

File: core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+    public <T> T invokeAny(Collection<? extends Callable<T>> tasks)
+        throws InterruptedException, ExecutionException {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <T> T invokeAny(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit)
+        throws InterruptedException, ExecutionException, TimeoutException {
+      throw new UnsupportedOperationException();
+    }
```
Comment (zeroFruit) [inline core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java:171]: It would be great if all of these unsupported methods are implemented in `DefaultEventLoopGroup` not in mock

File: core/src/main/java/io/el/concurrent/SingleThreadEventLoop.java [hunk 16]
```diff
@@ -151,7 +152,7 @@ private void start() {
       success = true;
     } finally {
       if (!success) {
-        stateUpdater.compareAndSet(this, State.STARTED, State.NOT_STARTED);
+        stateUpdater.compareAndSet(this, State.STARTED, State.TERMINATED);
```
Comment (zeroFruit) [inline core/src/main/java/io/el/concurrent/SingleThreadEventLoop.java:155]: If event loop is failed to start it is not in `TERMINATED` state but in `NOT_STARTED` state. plz revert this change

File: core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+@SuppressWarnings("unchecked")
+public class DefaultEventLoopGroupTest {
+
+  private static final SingleThreadEventLoop eventLoop = new MockEventLoop(
+      new ThreadPerTaskExecutor(Executors.defaultThreadFactory())
+  );
```
Comment (zeroFruit) [inline core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java:33]: It would be better just `Mockito.mock(SingleThreadEventLoop.class)` instead of `MockEventLoop`.

And I think `MockEventLoop` class you created is not needed

File: core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java (hunk 13 above)
Comment (raspLarry) [reply core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java:218]: Yes. I ll change the displayName contents.

File: core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java (hunk 14 above)
Comment (raspLarry) [reply core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java:198]: oh I've tried not to add exceptions on method signature, but I found that I already added exceptions on abstract newChild() method in parent. I'll fix it.

File: core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java (hunk 15 above)
Comment (raspLarry) [reply core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java:171]: done in oral discussion.

File: core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java (hunk 17 above)
Comment (raspLarry) [reply core/src/test/java/io/el/concurrent/DefaultEventLoopGroupTest.java]: done in oral discussion.
decided to delete DefaultEventLoopGroupTest.java

File: .gitignore [hunk 18]
```diff
@@ -1,4 +1,5 @@
 .idea/
 **/.DS_Store
 .gradle/
-**/build/
\ No newline at end of file
+**/build/
+core/out/**
```
Comment (zeroFruit) [inline .gitignore:5]: I think `**/out/**` would be more general than `core/out/**`

**Review by zeroFruit — APPROVED** [review]

**Review by majecty — APPROVED** [review]
